presto-common/src/main/java/com/facebook/presto/common/block/AbstractArrayBlock.java

```diff
@@ -51,35 +55,55 @@ public String getEncodingName()
     }
 
     @Override
-    public Block copyPositions(int[] positions, int offset, int length)
+    public final Block copyPositions(int[] positions, int offset, int length)
     {
         checkArrayRange(positions, offset, length);
 
+        int oldOffsetBase = getOffsetBase();
+        int[] oldOffsets = getOffsets();
         int[] newOffsets = new int[length + 1];
-        boolean[] newValueIsNull = new boolean[length];
-
-        IntArrayList valuesPositions = new IntArrayList();
-        int newPosition = 0;
-        for (int i = offset; i < offset + length; ++i) {
-            int position = positions[i];
-            if (isNull(position)) {
-                newValueIsNull[newPosition] = true;
-                newOffsets[newPosition + 1] = newOffsets[newPosition];
+        boolean[] newValueIsNull = mayHaveNull() ? new boolean[length] : null;
+        int currentOffset = 0;
+        // First traversal will populate newValueIsNull mask (if present) and
+        // newOffsets to determine the total number of value positions involved
+        for (int i = 0; i < length; i++) {
+            newOffsets[i] = currentOffset;
+            int position = positions[i + offset];
+            if (newValueIsNull != null && isNull(position)) {
+                newValueIsNull[i] = true;
             }
             else {
-                int valuesStartOffset = getOffset(position);
-                int valuesEndOffset = getOffset(position + 1);
+                int valuesEndOffset = oldOffsets[position + oldOffsetBase + 1];
+                int valuesStartOffset = oldOffsets[position + oldOffsetBase];
                 int valuesLength = valuesEndOffset - valuesStartOffset;
```

Comment:
The `valuesEndOffset` and `valuesStartOffset` are both just used once. Let's just inline them.